File: android/src/main/kotlin/com/koraidv/flutter/KoraIDVFlutterPlugin.kt
package com.koraidv.flutter

import android.app.Activity
import android.content.Intent
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.PluginRegistry
import com.koraidv.sdk.Configuration
import com.koraidv.sdk.DocumentType
import com.koraidv.sdk.KoraException
import com.koraidv.sdk.KoraIDV
import com.koraidv.sdk.KoraTheme
import com.koraidv.sdk.LivenessMode
import com.koraidv.sdk.Verification
import com.koraidv.sdk.VerificationRequest
import com.koraidv.sdk.VerificationTier
import com.koraidv.sdk.ui.VerificationActivity

/**
 * Flutter plugin for KoraIDV Android SDK.
 *
 * Implements [FlutterPlugin] + [ActivityAware] to get the current Activity
 * for launching the native verification UI via startActivityForResult.
 *
 * Exposes three MethodChannel calls:
 *   - configure(config Map)
 *   - startVerification(args Map) -> result Map
 *   - resumeVerification(args Map) -> result Map
 *
 * Mirrors koraidv-react-native's KoraIDVReactNativeModule.kt adapted
 * for Flutter's plugin architecture.
 */
class KoraIDVFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
    PluginRegistry.ActivityResultListener {

    companion object {
        private const val CHANNEL_NAME = "com.koraidv.flutter/koraidv"
        private const val REQUEST_CODE_VERIFICATION = 9002
    }

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var pendingResult: MethodChannel.Result? = null

    // -----------------------------------------------------------------------
    // FlutterPlugin
    // -----------------------------------------------------------------------

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    // -----------------------------------------------------------------------
    // ActivityAware
    // -----------------------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityBinding?.removeActivityResultListener(this)
        activity = null
        activityBinding = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activityBinding?.removeActivityResultListener(this)
        activity = null
        activityBinding = null
    }

    // -----------------------------------------------------------------------
    // MethodCallHandler
    // -----------------------------------------------------------------------

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "configure" -> handleConfigure(call, result)
            "startVerification" -> handleStartVerification(call, result)
            "resumeVerification" -> handleResumeVerification(call, result)
            else -> result.notImplemented()
        }
    }

    // -----------------------------------------------------------------------
    // configure
    // -----------------------------------------------------------------------

    private fun handleConfigure(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        if (args == null) {
            result.error("VALIDATION_ERROR", "Missing configuration arguments.", null)
            return
        }

        val apiKey = args["apiKey"] as? String
        val tenantId = args["tenantId"] as? String
        if (apiKey == null || tenantId == null) {
            result.error("INVALID_API_KEY", "Missing apiKey or tenantId.", null)
            return
        }

        val documentTypes = (args["documentTypes"] as? List<*>)?.mapNotNull { code ->
            DocumentType.fromCode(code as? String ?: "")
        } ?: DocumentType.entries

        val livenessMode = if ((args["livenessMode"] as? String) == "passive") {
            LivenessMode.PASSIVE
        } else {
            LivenessMode.ACTIVE
        }

        val theme = (args["theme"] as? Map<*, *>)?.let { themeMap ->
            KoraTheme(
                primaryColor = (themeMap["primaryColor"] as? String)?.hexToLong()
                    ?: KoraTheme().primaryColor,
                backgroundColor = (themeMap["backgroundColor"] as? String)?.hexToLong()
                    ?: KoraTheme().backgroundColor,
                surfaceColor = (themeMap["surfaceColor"] as? String)?.hexToLong()
                    ?: KoraTheme().surfaceColor,
                textColor = (themeMap["textColor"] as? String)?.hexToLong()
                    ?: KoraTheme().textColor,
                errorColor = (themeMap["errorColor"] as? String)?.hexToLong()
                    ?: KoraTheme().errorColor,
                successColor = (themeMap["successColor"] as? String)?.hexToLong()
                    ?: KoraTheme().successColor,
                cornerRadius = (themeMap["cornerRadius"] as? Double)?.toFloat() ?: 12f
            )
        } ?: KoraTheme()

        val config = Configuration(
            apiKey = apiKey,
            tenantId = tenantId,
            baseUrl = (args["baseUrl"] as? String)?.ifEmpty { null },
            documentTypes = documentTypes,
            livenessMode = livenessMode,
            theme = theme,
            timeout = (args["timeout"] as? Number)?.toLong() ?: 600L,
            debugLogging = (args["debugLogging"] as? Boolean) ?: false
        )

        KoraIDV.configure(config)
        result.success(null)
    }

    // -----------------------------------------------------------------------
    // startVerification
    // -----------------------------------------------------------------------

    private fun handleStartVerification(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val externalId = args?.get("externalId") as? String
        val tierString = args?.get("tier") as? String

        if (externalId == null || tierString == null) {
            result.error("VALIDATION_ERROR", "Missing externalId or tier.", null)
            return
        }

        val currentActivity = activity
        if (currentActivity == null) {
            result.error("NOT_CONFIGURED", "No current activity found.", null)
            return
        }

        if (!KoraIDV.isConfigured) {
            result.error("NOT_CONFIGURED", "SDK not configured. Call KoraIDV.configure() first.", null)
            return
        }

        val tier = when (tierString) {
            "basic" -> VerificationTier.BASIC
            "enhanced" -> VerificationTier.ENHANCED
            else -> VerificationTier.STANDARD
        }

        val request = VerificationRequest(
            externalId = externalId,
            tier = tier
        )

        pendingResult = result

        val intent = VerificationActivity.createIntent(currentActivity, request)
        currentActivity.startActivityForResult(intent, REQUEST_CODE_VERIFICATION)
    }

    // -----------------------------------------------------------------------
    // resumeVerification
    // -----------------------------------------------------------------------

    private fun handleResumeVerification(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val verificationId = args?.get("verificationId") as? String

        if (verificationId == null) {
            result.error("VALIDATION_ERROR", "Missing verificationId.", null)
            return
        }

        val currentActivity = activity
        if (currentActivity == null) {
            result.error("NOT_CONFIGURED", "No current activity found.", null)
            return
        }

        if (!KoraIDV.isConfigured) {
            result.error("NOT_CONFIGURED", "SDK not configured. Call KoraIDV.configure() first.", null)
            return
        }

        pendingResult = result

        val intent = VerificationActivity.createResumeIntent(currentActivity, verificationId)
        currentActivity.startActivityForResult(intent, REQUEST_CODE_VERIFICATION)
    }

    // -----------------------------------------------------------------------
    // Activity result
    // -----------------------------------------------------------------------

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode != REQUEST_CODE_VERIFICATION) return false

        val result = pendingResult ?: return false
        pendingResult = null

        when (resultCode) {
            Activity.RESULT_OK -> {
                @Suppress("DEPRECATION")
                val verification = data?.getParcelableExtra<Verification>(
                    VerificationActivity.EXTRA_VERIFICATION
                )
                if (verification != null) {
                    result.success(Serialization.buildSuccessResult(verification))
                } else {
                    result.error("UNKNOWN", "Missing verification data in result.", null)
                }
            }
            Activity.RESULT_CANCELED -> {
                @Suppress("DEPRECATION")
                val error = data?.getParcelableExtra<KoraException>(
                    VerificationActivity.EXTRA_ERROR
                )
                if (error != null) {
                    result.error(error.errorCode, error.message, null)
                } else {
                    result.success(Serialization.buildCancelledResult())
                }
            }
            else -> {
                result.success(Serialization.buildCancelledResult())
            }
        }

        return true
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private fun String.hexToLong(): Long? {
        if (this.isEmpty()) return null
        val hex = this.removePrefix("#")
        if (hex.length != 6) return null
        return try {
            (0xFF000000 or hex.toLong(16))
        } catch (_: NumberFormatException) {
            null
        }
    }
}

File: ios/Classes/KoraIDVFlutterPlugin.swift
import Flutter
import UIKit
import SwiftUI
import KoraIDV

/// Flutter plugin for KoraIDV iOS SDK.
///
/// Exposes three MethodChannel calls:
///   - configure(config Map)
///   - startVerification(args Map) -> result Map
///   - resumeVerification(args Map) -> result Map
///
/// Results cross the bridge as [String: Any] dictionaries via
/// Flutter's StandardMethodCodec (no JSON string encoding needed).
///
/// Mirrors the RN bridge (KoraIDVReactNative.swift) adapted for Flutter.
public class KoraIDVFlutterPlugin: NSObject, FlutterPlugin {

    // MARK: - Plugin registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.koraidv.flutter/koraidv",
            binaryMessenger: registrar.messenger()
        )
        let instance = KoraIDVFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // MARK: - Method call handler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":
            handleConfigure(call, result: result)
        case "startVerification":
            handleStartVerification(call, result: result)
        case "resumeVerification":
            handleResumeVerification(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - configure

    private func handleConfigure(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let apiKey = args["apiKey"] as? String,
              let tenantId = args["tenantId"] as? String
        else {
            result(FlutterError(code: "INVALID_API_KEY", message: "Missing apiKey or tenantId.", details: nil))
            return
        }

        var config = Configuration(apiKey: apiKey, tenantId: tenantId)

        if let env = args["environment"] as? String {
            config.environment = env == "sandbox" ? .sandbox : .production
        }

        if let baseUrl = args["baseUrl"] as? String, let url = URL(string: baseUrl) {
            config.baseURL = url
        }

        if let docTypes = args["documentTypes"] as? [String] {
            config.documentTypes = docTypes.compactMap { DocumentType(rawValue: $0) }
        }

        if let liveness = args["livenessMode"] as? String {
            config.livenessMode = liveness == "passive" ? .passive : .active
        }

        if let themeMap = args["theme"] as? [String: Any] {
            var theme = KoraTheme()
            if let primary = themeMap["primaryColor"] as? String {
                theme.primaryColor = Color(hex: primary)
            }
            if let bg = themeMap["backgroundColor"] as? String {
                theme.backgroundColor = Color(hex: bg)
            }
            if let surface = themeMap["surfaceColor"] as? String {
                theme.surfaceColor = Color(hex: surface)
            }
            if let text = themeMap["textColor"] as? String {
                theme.textColor = Color(hex: text)
            }
            if let error = themeMap["errorColor"] as? String {
                theme.errorColor = Color(hex: error)
            }
            if let success = themeMap["successColor"] as? String {
                theme.successColor = Color(hex: success)
            }
            if let radius = themeMap["cornerRadius"] as? CGFloat {
                theme.cornerRadius = radius
            }
            config.theme = theme
        }

        if let timeout = args["timeout"] as? TimeInterval {
            config.timeout = timeout
        }

        if let debug = args["debugLogging"] as? Bool {
            config.debugLogging = debug
        }

        KoraIDV.configure(with: config)
        result(nil)
    }

    // MARK: - startVerification

    private func handleStartVerification(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let externalId = args["externalId"] as? String,
              let tierString = args["tier"] as? String
        else {
            result(FlutterError(code: "VALIDATION_ERROR", message: "Missing externalId or tier.", details: nil))
            return
        }

        let tier: VerificationTier
        switch tierString {
        case "basic": tier = .basic
        case "enhanced": tier = .enhanced
        default: tier = .standard
        }

        let expectedFirstName = args["expectedFirstName"] as? String
        let expectedLastName = args["expectedLastName"] as? String

        DispatchQueue.main.async {
            guard let presenter = Self.topViewController() else {
                result(FlutterError(code: "NOT_CONFIGURED", message: "No presenting view controller found.", details: nil))
                return
            }

            KoraIDV.startVerification(
                externalId: externalId,
                tier: tier,
                expectedFirstName: expectedFirstName,
                expectedLastName: expectedLastName,
                from: presenter
            ) { verificationResult in
                Self.handleVerificationResult(verificationResult, result: result)
            }
        }
    }

    // MARK: - resumeVerification

    private func handleResumeVerification(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let verificationId = args["verificationId"] as? String
        else {
            result(FlutterError(code: "VALIDATION_ERROR", message: "Missing verificationId.", details: nil))
            return
        }

        DispatchQueue.main.async {
            guard let presenter = Self.topViewController() else {
                result(FlutterError(code: "NOT_CONFIGURED", message: "No presenting view controller found.", details: nil))
                return
            }

            KoraIDV.resumeVerification(
                verificationId: verificationId,
                from: presenter
            ) { verificationResult in
                Self.handleVerificationResult(verificationResult, result: result)
            }
        }
    }

    // MARK: - Result handling

    private static func handleVerificationResult(
        _ verificationResult: VerificationResult,
        result: @escaping FlutterResult
    ) {
        switch verificationResult {
        case .success(let verification):
            result(FlutterSerialization.buildSuccessResult(verification))

        case .failure(let error):
            result(FlutterError(code: error.errorCode, message: error.message, details: nil))

        case .cancelled:
            result(FlutterSerialization.buildCancelledResult())
        }
    }

    // MARK: - Top view controller

    private static func topViewController() -> UIViewController? {
        guard let window = UIApplication.shared.delegate?.window ?? nil,
              var vc = window.rootViewController
        else {
            return nil
        }
        while let presented = vc.presentedViewController {
            vc = presented
        }
        return vc
    }
}

